sstables_manager: do not reclaim unlinked sstables

lkshminarayanan opened the pull request:
When an sstable is unlinked, it remains in the _active list of the
sstable manager. Its memory might be reclaimed and later reloaded,
causing issues since the sstable is already unlinked. This patch updates
the on_unlink method to reclaim memory from the sstable upon unlinking,
remove it from memory tracking, and thereby prevent the issues described
above.

Added a testcase to verify the fix.

Fixes #21887

This is a bug fix in the bloom filter reload/reclaim mechanism and should be backported to older versions.

raphaelsc commented on sstables/sstables.hh:
```diff
@@ -142,6 +142,10 @@ inline std::string_view state_to_dir(sstable_state state) {
+        // this should never be called and is here for completeness
+        SCYLLA_ASSERT(false);
+        return "";
     }
```
I'd rather not introduce a unlinked to sstable state, since as you can see, it's a bit more specific than a generic state. And I am sure we can break this assumption (e.g. compaction racing with streaming) :-) 

I'd prefer if we could confine the unlinked tracking in manager, by introducing a 'set_type _unlinked', which we update in on_unlink() and deactivate().

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13766/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13766/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/bloom_filter_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/637/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/39/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2216/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13343/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5334/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13766/)
#### Failed Tests (1/36178):
- [test_view_build_status_with_replace_node.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13766/testReport/junit/topology_custom/test_view_build_status/Tests___Unit_Tests___test_view_build_status_with_replace_node_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_build_status_with_replace_node.dev.2)
#### Build Details:
- Duration: 8 hr 8 min
- Builder: i-0915f81b1b5f7417d (m5ad.8xlarge)

denesb commented on sstables/sstables_manager.cc:
```diff
@@ -196,6 +196,15 @@ future<> sstables_manager::components_reloader_fiber() {
+                // This sstable has been unlinked already,
+                // remove it from reclaimed and try the next sstable.
+                _reclaimed.erase(sstable_to_reload);
+                _total_memory_reclaimed -= reclaimed_memory;
```
Can we just add `_total_memory_reclaimed = 0` to `sstable::unlink()` so it is not even considered by this loop instead?

lkshminarayanan replied:
@denesb realised after typing the above comment - I can just free the memory in unlink() and set the  `_total_memory_reclaimed` to 0. That will fix the issue.

lkshminarayanan replied:
The patch sets the _total_memory_reclaimed on unlink now.

lkshminarayanan commented:
Changelog V2:
 - On unlink, the _total_memory_reclaimed is set to 0 and thus the unlinked sstable avoids the reclaim/reload cycle despite being in the _active list.
 - Removed the old logic

lkshminarayanan commented:
> #### Failed Tests (1/36178):
> * [test_view_build_status_with_replace_node.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13766/testReport/junit/topology_custom/test_view_build_status/Tests___Unit_Tests___test_view_build_status_with_replace_node_dev_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_build_status_with_replace_node.dev.2)

https://github.com/scylladb/scylladb/issues/20638

denesb commented on sstables/sstables_manager.cc:
```diff
@@ -330,6 +330,9 @@ std::vector<std::filesystem::path> sstables_manager::get_local_directories(const
 void sstables_manager::on_unlink(sstable* sst) {
+    // Reclaim memory used by the sstable components and update the tracker
+    auto memory_reclaimed = sst->reclaim_memory_from_components();
+    _total_reclaimable_memory -= memory_reclaimed;
```
I think we should just 0 both `_total_reclaimable_memory` and `_total_memory_reclaimed` here, just to be on the safe side.

raphaelsc replied:
this is in sstables_manager, so global stats. @lkshminarayanan don't we have to adjust _total_memory_reclaimed here? since it's decreased in deactive().


I think we can extract
```
    auto memory_reclaimed = sst_with_max_memory->reclaim_memory_from_components();
    _total_memory_reclaimed += memory_reclaimed;
    _total_reclaimable_memory -= memory_reclaimed;
```
into a function and reuse it here.

lkshminarayanan replied:
you're right. I'll update the patch.

lkshminarayanan commented:
changelog v3:
 - fix updation of `_total_memory_reclaimed`

denesb commented on sstables/sstables_manager.cc:
```diff
@@ -334,6 +334,8 @@ std::vector<std::filesystem::path> sstables_manager::get_local_directories(const
 
 void sstables_manager::on_unlink(sstable* sst) {
+    // Reclaim memory used by the sstable components
+    reclaim_memory_from_sstable(*sst);
```
How does this prevent the sstable from being selected for reload?

lkshminarayanan replied:
The sstable with the highest memory usage is reclaimed whenever a new sstable is added, and the total component memory, including the new sstable's memory, exceeds the threshold. This guarantees that whenever a reclamation code is triggered, there exists atleast one sstable whose component memory is non zero, ensuring that any sstable with components memory == 0 will never be picked up for reclamation.

Relevent code here - https://github.com/scylladb/scylladb/blob/98860905d820efa15458458e9c93f148fd81b62f/sstables/sstables_manager.cc#L159-L168

lkshminarayanan replied:
Shall I add the following line after `auto sst_with_max_memory = std::max_element...`?
```
    if (sst_with_max_memory->total_reclaimable_memory_size() == 0) {
        return;
    }
```

As per the logic I explained above this `if` will never be true but it will make the code more clear.

denesb replied:
I ask about the other case: when we have some free memory and want to reload some previously evicted sstable component. How will the above prevent an unlinked sstable from being picked for reloading its component?

lkshminarayanan replied:
Only sstables from `_reclaimed` set will be picked up for reload. Unlinked table will never end up in that set for two reasons 
1. As explained above they will never be picked up for reclaim and hence never be reloaded
2. If they were already in `_reclaimed` when they are unlinked, we explicitly remove them from that set. (#19722)

denesb replied:
I see, I though the bug we fix here is that we unlink sstables that are already reclaimed, but after re-reading the issue, I realized the bug is that we reclaim already unlinked sstables (then reload them later).

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13788/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13788/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/bloom_filter_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/656/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2257/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/58/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5376/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13383/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13788/)
#### Failed Tests (1/36162):
- [cqlpy.cassandra_tests/validation/entities/static_columns_test.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13788/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_static_columns_test_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/static_columns_test.debug.3)
#### Build Details:
- Duration: 8 hr 11 min
- Builder: i-088083e45fd8f9cfd (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13795/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13795/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/bloom_filter_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2264/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/65/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5383/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/663/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13393/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13795/)
#### Failed Tests (3/36156):
- [test_rolling_cluster_replacement_sequentially_live_nodes_multi_dc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13795/testReport/junit/cluster_replacement_test/TestClusterReplacement/Tests___dtest___test_rolling_cluster_replacement_sequentially_live_nodes_multi_dc) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_cluster_replacement_sequentially_live_nodes_multi_dc)
- [test_bloom_filter[500-1000-0.1].dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13795/testReport/junit/cqlpy/test_bloom_filter/Tests___Unit_Tests___test_bloom_filter_500_1000_0_1__dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_bloom_filter[500-1000-0.1].dev.3)
- [cqlpy.test_bloom_filter.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13795/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_test_bloom_filter_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.test_bloom_filter.dev.3)
#### Build Details:
- Duration: 4 hr 15 min
- Builder: spider1.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13872/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13872/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/bloom_filter_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/725/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5446/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2327/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13455/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/127/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13872/)
#### Failed Tests (2/36273):
- [test_slow_query_logging](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13872/testReport/junit/alternator_tests/TesterAlternator/Tests___dtest_with_tablets___test_slow_query_logging) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_slow_query_logging)
- [test_index_doesnt_flood_cache_in_small_partition_workload](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13872/testReport/junit/boost/cache_algorithm_test/test_index_doesnt_flood_cache_in_small_partition_workload) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_index_doesnt_flood_cache_in_small_partition_workload)
#### Build Details:
- Duration: 4 hr 6 min
- Builder: spider2.cloudius-systems.com

denesb commented:
Very unlucky PR :(
I re-kicked CI again, @lkshminarayanan please annotate previous CI failures.

denesb commented on sstables/sstables.cc:
```diff
@@ -1527,6 +1527,7 @@ size_t sstable::reclaim_memory_from_components() {
             // No need to remove it from _recognized_components as the filter is still in disk.
             _components->filter = std::make_unique<utils::filter::always_present_filter>();
             memory_reclaimed_this_iteration += filter_memory_size;
+            sstlog.info("Reclaimed bloom filter of {}", get_filename());
```
Isn't this redundant with the log that exists in `reclaim_memory_from_sstable()`?

lkshminarayanan replied:
I have removed it in the new version.

denesb commented on sstables/sstables_manager.cc:
```diff
@@ -152,6 +152,13 @@ sstable_writer_config sstables_manager::configure_writer(sstring origin) const {
+    auto memory_reclaimed = sst.reclaim_memory_from_components();
+    _total_memory_reclaimed += memory_reclaimed;
+    _total_reclaimable_memory -= memory_reclaimed;
+    smlogger.info("Reclaimed {} bytes of memory from SSTable components. Total memory reclaimed so far is {} bytes", memory_reclaimed, _total_memory_reclaimed);
```
We can make `reclaim_memory_from_components()` return the list of evicted components (in addition to the current return value) and then you can also add that to the log line here, and drop the previous patch.

lkshminarayanan replied:
I have updated the log here to print the sstable file name. For now we only reclaim bloom filters so, I have not `reclaim_memory_from_components` to return all the components that are being reclaimed. If/When we add support to reclaim more components, we can update the code to return an exact list.

lkshminarayanan commented:
> Very unlucky PR :( I re-kicked CI again, @lkshminarayanan please annotate previous CI failures.

The PR is not unlucky rather incomplete. :(
The previous issues were all related to bloom filters. I'm not able to reproduce them locally though. 
Something is still missing in this PR - possibly related to what you mentioned here https://github.com/scylladb/scylladb/pull/21895#discussion_r1882158062

denesb commented:
Cover-letter needs update.

lkshminarayanan commented:
Changelog V4: Another attempt to fix the issue with the following changes :
- Implemented `disable_component_memory_reload()` to sstables class that just sets the `_total_memory_reclaimed` of an sstable to 0. This effectively disables reload after that point.
- Implemented `reclaim_memory_and_stop_tracking_sstable()` to reclaim memory from an sstable (if it has not been done already) and remove the sstable from reclaimable tracking metrics. This is now called during both deactivate() and unlink() and post this call, any attempts by the sstable manager to reload or reclaim from the sstable will be a no-op. The method is also idempotent, so calling it twice from deactivate() and unlink() is safe. 
- Updated the testcases

lkshminarayanan commented:
@raphaelsc @denesb This is now ready for review.

denesb commented on sstables/sstables_manager.cc:
```diff
@@ -171,7 +171,9 @@ void sstables_manager::increment_total_reclaimable_memory_and_maybe_reclaim(ssta
-    smlogger.info("Reclaimed {} bytes of memory from SSTable components. Total memory reclaimed so far is {} bytes", memory_reclaimed, _total_memory_reclaimed);
+    // TODO: As of now only bloom filter is reclaimed. Print actual component names when adding support for more components.
+    smlogger.info("Reclaimed {} bytes of memory from components of {}. Total memory reclaimed so far is {} bytes",
+            sst_with_max_memory->get_filename(), memory_reclaimed, _total_memory_reclaimed);
```
Won't this put the sstable's path in the `Reclaimed {} bytes`?

lkshminarayanan replied:
my bad. I'll update it.

lkshminarayanan replied:
fixed it now.

denesb reviewed (approved):
LGTM, one small comment.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13901/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13901/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/bloom_filter_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/749/)
:x: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/151/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13486/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2357/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5476/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13901/)
#### Failed Tests (1/36193):
- [test_mv_topology_change.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13901/testReport/junit/topology_custom/test_mv_topology_change/Tests___Unit_Tests___test_mv_topology_change_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_topology_change.dev.3)
#### Build Details:
- Duration: 8 hr 9 min
- Builder: i-012b2a6305130c456 (m5ad.8xlarge)

lkshminarayanan commented:
> #### Failed Tests (1/36193):
> * [test_mv_topology_change.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13901/testReport/junit/topology_custom/test_mv_topology_change/Tests___Unit_Tests___test_mv_topology_change_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_topology_change.dev.3)

From the [test_py.log](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13901/artifact/testlog/x86_64/dev/failed_test/test_mv_topology_change.3.dev.3/test_py.log) :
```
21:06:39.487 INFO> [dev/topology_custom.test_mv_topology_change.1] Cluster 0b788170-bcaa-11ef-8fe2-0e0ae25cfe27 waiting for new IP...
21:06:39.487 INFO> [dev/topology_custom.test_mv_topology_change.1] Cluster 0b788170-bcaa-11ef-8fe2-0e0ae25cfe27 obtained new IP: 127.22.148.44
21:06:39.490 INFO> [dev/topology_custom.test_mv_topology_change.1] Cluster ScyllaCluster(name: 0b788170-bcaa-11ef-8fe2-0e0ae25cfe27, running: , stopped: ) adding server...
21:06:39.495 INFO> [dev/topology_custom.test_mv_topology_change.1] installing Scylla server in /scylladir/testlog/x86_64/dev/scylla-3946...
21:07:33.026 INFO> [dev/topology_custom.test_mv_topology_change.3] [ScyllaClusterManager][Task-253755] http://api/cluster/all-servers
21:07:33.028 INFO>  [17/Dec/2024:21:07:33 +0200] "GET /cluster/all-servers HTTP/1.1" 200 303 "-" "Python/3.12 aiohttp/3.9.5"
21:07:33.031 INFO> [dev/topology_custom.test_mv_topology_change.3] [ScyllaClusterManager][Task-253756] http://api/cluster/server/3940/get_log_filename
```

There is no update after `installing Scylla server...` and the `server_add` call times out.

lkshminarayanan commented:
> > #### Failed Tests (1/36193):
> > 
> > * [test_mv_topology_change.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13901/testReport/junit/topology_custom/test_mv_topology_change/Tests___Unit_Tests___test_mv_topology_change_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_topology_change.dev.3)
> 
> From the [test_py.log](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13901/artifact/testlog/x86_64/dev/failed_test/test_mv_topology_change.3.dev.3/test_py.log) :
> 
> ```
> 21:06:39.487 INFO> [dev/topology_custom.test_mv_topology_change.1] Cluster 0b788170-bcaa-11ef-8fe2-0e0ae25cfe27 waiting for new IP...
> 21:06:39.487 INFO> [dev/topology_custom.test_mv_topology_change.1] Cluster 0b788170-bcaa-11ef-8fe2-0e0ae25cfe27 obtained new IP: 127.22.148.44
> 21:06:39.490 INFO> [dev/topology_custom.test_mv_topology_change.1] Cluster ScyllaCluster(name: 0b788170-bcaa-11ef-8fe2-0e0ae25cfe27, running: , stopped: ) adding server...
> 21:06:39.495 INFO> [dev/topology_custom.test_mv_topology_change.1] installing Scylla server in /scylladir/testlog/x86_64/dev/scylla-3946...
> 21:07:33.026 INFO> [dev/topology_custom.test_mv_topology_change.3] [ScyllaClusterManager][Task-253755] http://api/cluster/all-servers
> 21:07:33.028 INFO>  [17/Dec/2024:21:07:33 +0200] "GET /cluster/all-servers HTTP/1.1" 200 303 "-" "Python/3.12 aiohttp/3.9.5"
> 21:07:33.031 INFO> [dev/topology_custom.test_mv_topology_change.3] [ScyllaClusterManager][Task-253756] http://api/cluster/server/3940/get_log_filename
> ```
> 
> There is no update after `installing Scylla server...` and the `server_add` call times out.

https://github.com/scylladb/scylladb/issues/21912

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13917/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13917/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/bloom_filter_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2373/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5490/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13502/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/164/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/762/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13917/)
#### Build Details:
- Duration: 4 hr 5 min
- Builder: spider7.cloudius-systems.com
